gst/nnstreamer/tensor_query/tensor_query_common.c

```diff
@@ -60,32 +57,422 @@ typedef struct
+      GSocketAddress *saddr;
+      client_socket = g_socket_accept (sdata->server_socket,
+          sdata->cancellable, &err);
+      if (!client_socket) {
+        nns_loge ("Failed to accept client: %s", err->message);
+        g_error_free (err);
+        return 0;
+      }
+      saddr = g_socket_get_remote_address (client_socket, &err);
+      if (!saddr) {
+        nns_loge ("Failed to get client address: %s", err->message);
+        g_error_free (err);
+        return 0;
+      }
+
+      conn = g_new0 (TensorQueryConnection, 1);
+      conn->protocol = sdata->protocol;
+      conn->host =
+          g_inet_address_to_string (g_inet_socket_address_get_address (
+              (GInetSocketAddress *) saddr));
+      conn->port =
+          g_inet_socket_address_get_port ((GInetSocketAddress *) saddr);
+      conn->socket = client_socket;
+      conn->cancellable = g_cancellable_new ();
+
+      nns_logi ("Inserted client %s:%d", conn->host, conn->port);
+      g_object_unref (saddr);
+    }
+      break;
+    default:
+      /* NYI */
+      return 0;
+  }
+  return conn;
+}
+
+/**
+ * @brief [TCP] establish socket for tcp server
+ */
+static GSocket *
+query_establish_socket (const char *host, uint32_t port)
+{
+  GSocket *server_socket;
+  GSocketAddress *saddr;
+  GError *err;
+
+  saddr = g_inet_socket_address_new_from_string (host, port);
+  if (!saddr) {
+    /* hostname resolution not supported */
+    nns_loge ("Failed to parse host `%s:%u`", host, port);
+    goto fail;
+  }
+
+  server_socket =
+      g_socket_new (g_socket_address_get_family (saddr), G_SOCKET_TYPE_STREAM,
+      G_SOCKET_PROTOCOL_TCP, &err);
+
+  if (!server_socket) {
+    nns_loge ("Failed to create socket : %s", err->message);
+    goto fail;
+  }
+
+  if (!g_socket_bind (server_socket, saddr, TRUE, &err)) {
+    nns_loge ("Failed to bind on host `%s:%u`: %s", host, port, err->message);
+    goto fail;
+  }
+
+  g_socket_set_listen_backlog (server_socket, TCP_BACKLOG);
+  /* listen */
+  if (!g_socket_listen (server_socket, &err)) {
+    nns_loge ("Failed to listen on host `%s:%u`: %s", host, port, err->message);
+    goto close_socket;
+  }
+  nns_logd ("Listening on port %d",
+      g_inet_socket_address_get_port ((GInetSocketAddress *) saddr));
+  g_object_unref (saddr);
+  return server_socket;
+
+fail:
+  g_object_unref (saddr);
+  g_error_free (err);
+  return NULL;
+
+close_socket:
+  if (!g_socket_close (server_socket, &err)) {
+    nns_loge ("Failed to close socket: %s", err->message);
+  }
+  g_object_unref (server_socket);
+  g_object_unref (saddr);
+  g_error_free (err);
+  return NULL;
+}
+
+/**
+ * @brief [TCP] receive data for tcp server
+ */
+static gboolean
+query_tcp_receive (GSocket * socket, uint8_t * data, size_t size,
+    GCancellable * cancellable)
+{
+  gsize bytes_received = 0;
+  gssize rret;
+  GError *err;
+
+  while (bytes_received < size) {
+    rret = g_socket_receive (socket, (gchar *) data + bytes_received,
```

Comment:
Is it safe to use mixed signed and unsigned data types here and at line 450?

Reply:
I think it 's ok as long as we use it for byte buffer.
For L450, it looks fine because size is added only if >= 0